### PR TITLE
update the unittests of the solver using a solver factory

### PR DESCRIPTION
**Pull request by MaximilienNaveau**
## What changed?

- creation of a solver factory for all the available solvers.
- update the test_solvers.cpp using the factory, this should remove the segfault from the gepgitlab build farm.

### CODE REVIEW

**Comment by cmastalli** [comment]
@MaximilienNaveau thanks for this nice job! do you know why the CI is failing?

**Comment by wxmerkt** [comment]
```
The following tests FAILED:

	  1 - test_state (Failed)

	  5 - test_impulses (Failed)

	 10 - py-actuation (Failed)

	 11 - py-armature (Failed)

	 13 - py-contacts (Failed)

	 14 - py-costs (Failed)

	 15 - py-ddp_contact (Failed)

	 16 - py-dse3 (Failed)

	 17 - py-dynamic_derivatives (Failed)

	 18 - py-impact (Failed)

	 19 - py-quadruped (Failed)

	 21 - py-robots (Failed)

	 23 - py-state (Failed)
```

**Comment by cmastalli** [comment]
@wxmerkt do you know why I am not able to trigger the CI jobs?

**Comment by wxmerkt** [comment]
> @wxmerkt do you know why I am not able to trigger the CI jobs?

You may need to click `Sign In with GitHub` on Travis. As a project owner you should be able to trigger after logging in.

**Comment by wxmerkt** [comment]
The tests are failing due to the change in example-robot-data and there not being a release yet.

**Comment by MaximilienNaveau** [comment]
> @MaximilienNaveau thanks for this nice job! do you know why the CI is failing?

I think it is because you need to create all the memory inside the unit-test function. When you declare the tests not memory should be instantiated or just some int or double that you copy.

**Comment by MaximilienNaveau** [comment]
> The tests are failing due to the change in example-robot-data and there not being a release yet.

Is this related to: https://github.com/loco-3d/crocoddyl/pull/607 ?

**Comment by MaximilienNaveau** [comment]
How can I check the MPI/Travis CI with the new setup?

File: unittest/solver_factory.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2018-2019, LAAS-CNRS
```
[the remark this answers is not in the document]
Comment (MaximilienNaveau) [reply unittest/solver_factory.hpp]: ok, good point, should we add the same header in all files?

[the remark this answers is not in the document]
Comment (cmastalli) [reply unittest/solver_factory.hpp]: Yes, you should :)

**Comment by wxmerkt** [comment]
> How can I check the MPI/Travis CI with the new setup?

Not sure I understand the question. For Travis you can click on the `Details` button next to the `x` which directs you to the build page. There you can click on the configuration and get the full log. Is that what you meant?

File: unittest/solver_factory.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (wxmerkt) [reply unittest/solver_factory.hpp]: > ok, good point, should we add the same header in all files?

I'd say only amend the ones touched at every change/

**Comment by MaximilienNaveau** [comment]
Thank @wxmerkt ! And the MPI gepgitlab CI? is it only build on the mirrored devel/master branches?

**Comment by wxmerkt** [comment]
We can switch to ROS-Industrial CI which simplifies the build configuration and allows local builds soon. Plus, it's supported.

**Comment by cmastalli** [comment]
> Thank @wxmerkt ! And the MPI gepgitlab CI? is it only build on the mirrored devel/master branches?

At the moment we do not have mirrored this repo in gepgitlab, but it is in the TODO list of @nim65s. You can still push code there to test it, and it is a good idea to do it since there are more jobs there.

File: unittest/solver_factory.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        break;
+    }
+  }
+
+  ~SolverFactory() {}
+
+  boost::shared_ptr<crocoddyl::SolverAbstract> get_solver() { return solver_; }
+
+ private:
+  size_t nb_running_models_;                              //!< This is the number of models in the shooting problem.
```
[the remark this answers is not in the document]
Comment (MaximilienNaveau) [reply unittest/solver_factory.hpp]: I do not think we need it as the compiler returns no warning.

File: unittest/solver_factory.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmastalli) [reply unittest/solver_factory.hpp]: Yes, I agree with @wxmerkt. I apologize if that was not clear.

[the remark this answers is not in the document]
Comment (MaximilienNaveau) [reply unittest/solver_factory.hpp]: Ok, but in the end you want a:
// Copyright (C) 2018-2020, LAAS-CNRS, New York University, Max Planck Gesellschaft
//                          University of Edinburgh
Everywhere? no?
If this is the case I will update this file and create another pull request that updates all headers.

**Comment by MaximilienNaveau** [comment]
I tried to push on my gepgitlab fork but the CI seemed down or de-activated for me... I am not sure what is the status.

> I think he refers to locally run the CI jobs. We were doing it on gepgitlab through a Docker. I agree with @MaximilienNaveau that this is useful and needed

This is not what I meant but yes, running the full CI for all job locally before pushing would be extremely beneficial. We could see the problems immediately.

File: unittest/solver_factory.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaximilienNaveau) [reply unittest/solver_factory.hpp]: Ok, thank you very much for the explanation!

**Comment by wxmerkt** [comment]
> This is not what I meant but yes, running the full CI for all job locally before pushing would be extremely beneficial. We could see the problems immediately.

The CI job only does `make`, `make install`, `make test`. I don't think the failure right now is CI related?

**Comment by MaximilienNaveau** [comment]
> The CI job only does `make`, `make install`, `make test`. I don't think the failure right now is CI related?

I agree it does the exact command you said but in different environment: Ubuntu 16.04/18.04 in release/debug, python2/3, etc.
So it would be nice to launch all those tests locally before a pull request so we are sure what is going on on different setup. Using a docker(or similar) of course is the recommended option.

**Comment by nim65s** [comment]
@MaximilienNaveau you can just push to your fork on gepgitlab, this will run the CI. And you can setup mirroring from your gepgitlab for to your github fork, to ease the PR opening after the CI run.

**Comment by cmastalli** [comment]
> @MaximilienNaveau you can push to your fork on gepgitlab, this will run the CI. And you can setup mirroring from your gepgitlab for to your github fork, to ease the PR opening after the CI run.

@nim65s I would like to setup the mirroring in my own fork, how can I do it?

File: unittest/solver_factory.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmastalli) [reply unittest/solver_factory.hpp]: > Ok, but in the end you want a:
> // Copyright (C) 2018-2020, LAAS-CNRS, New York University, Max Planck Gesellschaft
> // University of Edinburgh
> Everywhere? no?
> If this is the case I will update this file and create another pull request that updates all headers.

Yes, that order is fair :)

**Comment by nim65s** [comment]
here: https://gepgitlab.laas.fr/cmastall/crocoddyl/-/settings/repository ; in "Mirroring repositories", add `https://cmastalli@github.com/cmastalli/crocoddyl` as the repository URL, and then add either your github password, or a github token, or a ssh key.

**Comment by cmastalli** [comment]
Thanks @MaximilienNaveau. Very clear job :smiley:

**Comment by cmastalli** [comment]
> https://cmastalli@github.com/cmastalli/crocoddyl

Thanks for the explanation

**Comment by MaximilienNaveau** [comment]
It is all green ;)

**Pull request merged**